docs: type generation for graphql servers (#4376)
graphql/graphql-js: website/pages/docs/type-generation.mdx

functions return values that match their expected shapes.

However, code-first development has tradeoffs:

- You won't get automatic type definitions for your resolvers unless you generate
them manually or infer them through wrappers.
- Schema documentation, testing, and tool compatibility may require you to export
the schema to SDL first.

You can still use type generation tools like GraphQL Code Generator in a code-first setup.
You just need to convert your schema into SDL.

To export your schema:

```ts
import { printSchema } from 'graphql';
import { schema } from './schema';
import { writeFileSync } from 'fs';

writeFileSync('./schema.graphql', printSchema(schema));
```

Once you've written the SDL, you can treat the project like a schema-first project
for type generation.

## Schema-first development

In a schema-first workflow, your GraphQL schema is written in SDL, for example, `.graphql`
or `.gql` files. This serves as the source of truth for your server. This approach
emphasizes clarity because your schema is defined independently from your business logic.

Schema-first development pairs well with type generation because the schema is 

> serializable and can be directly used by tools like GraphQL Code Generator.

Reviewer: I wonder if we should link to GraphQL Code Generator website? https://the-guild.dev/graphql/codegen

```ts
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  schema: './schema.graphql',
  generates: {
    './src/generated/resolvers-types.ts': {
      plugins: ['typescript', 'typescript-resolvers'],
    },
  },
};
export default config;
```

To run the generator: 

```bash
npx graphql-codegen
```

This creates a set of resolver types like:

```ts
export type QueryResolvers<ContextType = any> = {
  user?: Resolver<User, any, ContextType, RequireFields<QueryUserArgs, 'id'>>;
};
```

These types ensure that the `user` resolver expects an `id` argument and returns a 
`User`, giving you confidence and autocomplete while implementing your server logic.

## Using generated types in your server

Once generated, you can use these types directly in your resolver map:

```ts
import { QueryResolvers } from './generated/resolvers-types';

export const queryResolvers: QueryResolvers = {
  user: (parent, args, context) => {
    return context.db.getUser(args.id);
  },
};
```

You can also extract shared `ContextType` and `Resolver` 
utility types from the generated file and apply them across your codebase.

## Generating operation types

> 
> In addition to resolver types, you can generate types for GraphQL operations such
> as queries, mutations, and fragments. This is especially useful for shared integration tests
> or client logic that needs to match the schema precisely.
> 
> We recommend using the GraphQL Code Generator Server Preset for generating operation types.
> The server preset generates both resolver and operation types, without needing to install
> or configure additional plugins.
> 
> Suppose you have a query in `./src/operations/getUser.graphql`: 
> 
> ```graphql
> query GetUser($id: ID!) {
>   user(id: $id) {
>     id
>     name
>   }
> }
> ```
> 
> Update your codegen config:
> 
> ```ts
> import type { CodegenConfig } from '@graphql-codegen/cli';
> 
> const config: CodegenConfig = {
>   schema: './schema.graphql',
>   documents: './src/operations/**/*.graphql',
>   generates: {
>     './src/generated/': {
>       preset: 'graphql-codegen-preset-server',
>     },
>   },
> };
> export default config;
> ```
> 
> This produces types like `GetUserQuery` and `GetUserQueryVariables`, which you can
> import into your client code or test files.
> 
> ## Typing resolvers manually
> 
> If you aren't ready to introduce type generation, you can still get partial type safety
> using `graphql-js` built-in types.
> 
> ```ts
> import { GraphQLFieldResolver } from 'graphql';
> 
> const myResolver: GraphQLFieldResolver<MyParent, MyContext> = (
>   parent,
>   args,
>   context,
>   info
> ) => {
>   // ...
> };
> ```
> 
> This pattern may be enough for small projects or static schemas, but it 
> can be hard to maintain and scale without automation.

Reviewer: ```suggestion
With a schema-first workflow, you can:

- Generate resolver type definitions and files that match your schema
- Generate operation types for client queries, integration tests, or internal tooling
- Detect breaking changes and unused types through schema diffing tools

## Generating resolver types

We recommend using the [Server Preset](https://www.npmjs.com/package/@eddeee888/gcg-typescript-resolver-files) for a managed workflow. It automatically generates types and files based on your schema without needing extra plugin setup:

- resolver types, including parent types, arguments, return values, and context.
- resolver files with types wired up, ready for your business logic
- resolver map and type definitions to plug into GraphQL servers

Example codegen config:

```ts filename="codegen.ts"
import type { CodegenConfig } from "@graphql-codegen/cli";
import { defineConfig } from "@eddeee888/gcg-typescript-resolver-files";

const config: CodegenConfig = {
  schema: "src/**/schema.graphql",
  generates: {
    "src/schema": defineConfig({
      resolverGeneration: "minimal",
    }),
  },
};
export default config;
```

Note: this setup expects your schema is split into modules to improve readability and maintainability:

```text
├── src/
│   ├── schema/
│   │   ├── base/
│   │   │   ├── schema.graphql
│   │   ├── user/
│   │   │   ├── schema.graphql
│   │   ├── book/
│   │   │   ├── schema.graphql
```

To run the generator:

```bash
npx graphql-codegen
```

This creates resolver files that like:

```ts filename="src/schema/resolvers/Query/user.ts"
import type { QueryResolvers } from "./../../../types.generated";
export const user: NonNullable<QueryResolvers["user"]> = async (
  _parent,
  _arg,
  _ctx,
) => {
  /* Implement Query.user resolver logic here */
};
```

The user query resolver is typed to ensure that the `user` resolver expects an `id` argument and returns a
`User`, giving you confidence and autocomplete while implementing your server logic, which may look like this:

```ts filename="src/schema/resolvers/Query/user.ts"
export const user: NonNullable<QueryResolvers["user"]> = async (
  parent,
  args,
  context,
) => {
  return context.db.getUser(args.id);
};
```

Check out the [official Server Preset guide](https://the-guild.dev/graphql/codegen/docs/guides/graphql-server-apollo-yoga-with-server-preset) to learn about its other features, including mappers convention and static analysis for runtime safety.

## Generating operation types

In addition to resolver types, you can generate types for GraphQL operations such as queries, mutations, and fragments. This is especially useful for shared integration tests or client logic that needs to match the schema precisely.

We recommend using the GraphQL Code Generator [Client Preset](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client) for a managed workflow:

- write operations with GraphQL syntax in the same file where it is used
- get type-safety when using the result

Example codegen config

```ts filename="codegen.ts"
import type { CodegenConfig } from "@graphql-codegen/cli";

const config: CodegenConfig = {
  schema: "src/**/schema.graphql",
  documents: ["src/**/*.ts"],
  ignoreNoDocuments: true,
  generates: {
    "./src/graphql/": {
      preset: "client",
      config: {
        documentMode: "string",
      },
    },
  },
};

export default config;
```

Now, run codegen in watch mode:

```shell
npx graphql-codegen --config codegen.ts --watch
```

Now, we can import the `graphql` function from `src/graphql/` to write GraphQL operations directly in your TypeScript files:

```ts filename="src/index.ts"
import { graphql } from "./graphql";

const UserQuery = graphql(`
  query User($id: ID!) {
    user(id: ID!) {
      id
      fullName
    }
  }
`);

const response = await fetch("https://graphql.org/graphql/", {
  method: "POST",
  headers: {
    "Content-Type": "application/json",
    Accept: "application/graphql-response+json",
  },
  body: JSON.stringify({
    query: UserQuery,
    variables: { id: "1" },
  }),
});

if (!response.ok) {
  throw new Error("Network response was not ok");
}

const result: ResultOf<typeof UserQuery> = await response.json();

console.log(result);
```

Check out Client Preset guides for popular framework and GraphQL clients:

- [Vanilla TypeScript](https://the-guild.dev/graphql/codegen/docs/guides/vanilla-typescript)
- [React Query](https://the-guild.dev/graphql/codegen/docs/guides/react-query)
- [React / Vue](https://the-guild.dev/graphql/codegen/docs/guides/react-vue)
```

managed workflow. It automatically generates types and files based on your schema without needing extra plugin setup.

The Server Preset generates:

- Resolver types, including parent types, arguments, return values, and context
- Resolver files with types wired up, ready for your business logic
- A resolver map and type definitions to plug into GraphQL servers

This setup expects your schema is split into modules to improve readability and maintainability. For example:

```pgsql
├── src/
│   ├── schema/
│   │   ├── base/
│   │   │   ├── schema.graphql
│   │   ├── user/
│   │   │   ├── schema.graphql
│   │   ├── book/
│   │   │   ├── schema.graphql
```

Here's an example `codegen.ts` file using the Server Preset:

```ts filename="codegen.ts"
import type { CodegenConfig } from "@graphql-codegen/cli";
import { defineConfig } from "@eddeee888/gcg-typescript-resolver-files";

const config: CodegenConfig = {
  schema: "src/**/schema.graphql",
  generates: {
    "src/schema": defineConfig({
      resolverGeneration: "minimal",
    }),
  },
};

export default config;
```

To generate the resolver types and files, run:

```bash
npx graphql-codegen
```

This creates resolver files like:

> ```ts

Reviewer: ```suggestion
```ts filename="src/schema/user/resolvers/Query/user.ts"
```

import type { QueryResolvers } from "./../../../types.generated";

export const user: NonNullable<QueryResolvers["user"]> = async (
  _parent,
  _arg,
  _ctx,
) => {
  // Implement Query.user resolver logic here
};
```

The user query resolver is typed to ensure that the user resolver expects an id argument and returns a
User, giving you confidence and autocomplete while implementing your server logic, which may look like this:

> ```ts

Reviewer: ```suggestion
```ts filename="src/schema/user/resolvers/Query/user.ts"
```

export const user: NonNullable<QueryResolvers["user"]> = async (
  parent,
  args,
  context,
) => {
  return context.db.getUser(args.id);
};
```

See the official [Server Preset guide](https://the-guild.dev/graphql/codegen/docs/guides/graphql-server-apollo-yoga-with-server-preset) to learn about its other features, including mappers convention and static analysis for runtime safety.

## Generating operation types

In addition to resolver types, you can generate types for GraphQL operations such as queries, mutations, and 
fragments. This is especially useful for shared integration tests or client logic that needs to match the schema 
precisely.

We recommend using the GraphQL Code Generator [Client Preset](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client) for a managed workflow:

- Write operations with GraphQL syntax in the same file where it is used
- Get type-safety when using the result

Here's an example configuration using the Client Preset:

> ```ts

Reviewer: ```suggestion
```ts filename="codegen.ts"
```

import type { CodegenConfig } from "@graphql-codegen/cli";

const config: CodegenConfig = {
  schema: "src/**/schema.graphql",
  documents: ["src/**/*.ts"],
  ignoreNoDocuments: true,
  generates: {
    "./src/graphql/": {
      preset: "client",
      config: {
        documentMode: "string",
      },
    },
  },
};

export default config;
```

To keep generated types up to date as you edit your code, run the generator in watch mode:

```bash
npx graphql-codegen --config codegen.ts --watch
```

Once generated, import the `graphql` function from `src/graphql/` to write GraphQL operations 
directly in your TypeScript files:

> ```ts

Reviewer: ```suggestion
```ts filename="src/index.ts"
```

- You want strong type safety across your server logic
- Your schema is defined separately in SDL files
- Your API surface is large, rapidly evolving, or used by multiple teams
- You rely on TypeScript for editor tooling, autocomplete, or static analysis

By generating types directly from your schema, you can avoid drift between schema
definitions and implementation logic.

## Code-first development

In a code-first workflow, the schema is constructed entirely in JavaScript or TypeScript
using `graphql-js` constructors like `GraphQLObjectType`, `GraphQLSchema`, and others.
This approach is flexible and lets you build your schema programmatically using native
language features.

If you're using this approach with TypeScript, you already get some built-in type safety
with the types exposed by `graphql-js`. For example, TypeScript can help ensure your resolver
functions return values that match their expected shapes.

However, code-first development has tradeoffs:

- You won't get automatic type definitions for your resolvers unless you generate
them manually or infer them through wrappers.
- Schema documentation, testing, and tool compatibility may require you to export

> the schema to SDL first.

Reviewer 2: It's quite common that newbies think that you can "export" to SDL and then run that SDL and you're running the schema; I think this is primarily people who misunderstand what GraphQL itself is. Rather than talking about this as "exporting the schema" which isn't really accurate, we should think of it more as a "description of the schema" since it omits the execution part.

```suggestion
- Schema documentation, testing, and tool compatibility may require you to provide
  a description of the schema in SDL first.
```

You can still use type generation tools like GraphQL Code Generator in a code-first setup.
You just need to convert your schema into SDL.

> To export your schema:

Reviewer 2: ```suggestion
To produce an SDL description of your schema:
```

```ts
import { printSchema } from 'graphql';
import { schema } from './schema';
import { writeFileSync } from 'fs';

writeFileSync('./schema.graphql', printSchema(schema));
```

Once you've written the SDL, you can treat the project like a schema-first project

> for type generation.

Reviewer 2: ```suggestion
Once you've written the SDL, you can treat the project like an SDL-first project
for type generation.
```

## Schema-first development

In a schema-first workflow, your GraphQL schema is written in SDL, for example, `.graphql`

> or `.gql` files. This serves as the source of truth for your server. This approach

Reviewer 2: ```suggestion
or `.gql` (discouraged) files. This serves as the source of truth for your server. This approach
```

emphasizes clarity because your schema is defined independently from your business logic.

Schema-first development pairs well with type generation because the schema is 
serializable and can be directly used by tools like [GraphQL Code Generator](https://the-guild.dev/graphql/codegen).

With a schema-first workflow, you can:

- Generate resolver type definitions and files that match your schema
- Generate operation types for client queries, integration tests, or internal tooling
- Detect breaking changes and unused types through schema diffing tools

## Generating resolver types

To get started, install the required packages:

```bash
npm install graphql @graphql-codegen/cli @eddeee888/gcg-typescript-resolver-files
```

> This scoped package is publish by a community maintainer and is widely used for GraphQL server

Reviewer 2: ```suggestion
This scoped package is published by a community maintainer and is widely used for GraphQL server
```
